pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -242,6 +243,132 @@ type loadIdx struct {
 	chunk       int
 }
 
+type rangeLoadIdx struct {
+	offset uint32
+	estLen int64
+	// seriesEntry and rangeEntry are indices for where to put the result,
+	// they are not actual refs of series or chunks
+	seriesEntry int
+	rangeEntry  int
+}
+
+type bucketChunksRangesReader struct {
```

Comment:
A suggestion to unit test this. We don't have to get mad, but a unit test for the basic case may be nice.

Instead of storing `*bucketBlock`, we keep (a) the partitioner (b) a reference to the function `chunkRangeReader()` (c) the logger (d) the block ID. In the test we override the partitioner and `chunkRangeRange()`. The `chunkRangeRange()` returns a simple reader that returns the sequence of bytes in the range 0-254 each time you read bytes from it. When asserting the actual chunks data read we assert that the bytes range fetched match the expected one (you can test it with very small chunk ranges, so that the total size of data read is <= 254 bytes).